### ISSUE TITLE
[no-extra-non-null-assertion] Problems with optional chaining

### ISSUE THREAD

**Issue by HolgerJeromin** [typescript-eslint/typescript-eslint#2732]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-extra-non-null-assertion": ["error"]
  }
}
```

```TS
function foo(callback: (this: void) => void): void {
    callback();
}

let myName: string | null = 'name';
foo(() => {
    let testObj: Record<string, string> = { name: 'value' };
    console.log(testObj?.[myName!]);
});
```

**Expected Result**
`myName` can be null in the view of TS.  TS2538	(TS) Type 'null' cannot be used as an index type.
So the null assertion is not extra.

**Actual Result**
(ESLint) @typescript-eslint/no-extra-non-null-assertion TypeScript Forbidden extra non-null assertion.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.6.0` |
| `@typescript-eslint/parser`        | `4.6.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `6.8.0` |

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/e9d2ab638b6767700b52797e74b814ea059beaae/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts#L35

The selector is wrong!
Should be a pretty easy fix

**Comment by ddubrava** [comment]
Hey! Can I try to fix this issue? I don't have any experience with `estree`, could you please share any docs about `optional, computed, etc.` properties? I've already researched some code and found that the `[computed = false]` for MemberExpression would fix the problem, all tests are passing.

**Comment by bradzacher** [comment]
Sure thing! Happy to accept a PR.

- docs on ESLint selectors - https://eslint.org/docs/developer-guide/selectors
- The types for `MemberExpression` (this is our "spec")
  https://github.com/typescript-eslint/typescript-eslint/blob/c8a4dadeab8a64fb4768deda8f65475435dd2cad/packages/types/src/ts-estree.ts#L414-L416
  https://github.com/typescript-eslint/typescript-eslint/blob/c8a4dadeab8a64fb4768deda8f65475435dd2cad/packages/types/src/ts-estree.ts#L646-L662
- estree spec - https://github.com/estree/estree/
    - I wouldn't bother with this, but it's there in case you're interested.
- AST explorer - https://astexplorer.net/
    - This is a great tool which will show you the AST for whatever code you put in.

LMK if you need any extra pointers.
